[ Test ] Creating test module TC_CADMIN_1_11

j-ororke opened the pull request:
- Test module created following matter-test-scripts task (https://github.com/project-chip/matter-test-scripts/issues/286): 
> - Open commissioning window on DUT twice using ECM then BCM [DUT - Commissionee] 
> - Commission DUT to TH_CR1 (can be skipped if done in a preceding test) 
> - TH_CR1 reads the BasicCommissioningInfo attribute from the General Commissioning cluster on EP0 and saves the MaxCumulativeFailsafeSeconds as timeout 
> - TH_CR1 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout 
> - TH_CR2 fully commissions the DUT
> - TH_CR1 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout 
> - TH_CR1 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout 
> - TH_CR2 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout 
> - TH_CR1 sends an RevokeCommissioning command to the DUT 
> - TH_CR1 reads the FeatureMap from the Administrator Commissioning Cluster. If the feature map includes the BC feature bit, repeat steps 5-8 using the OpenBasicCommissioningWindow command 
> - TH_CR2 reads the CurrentFabricIndex attribute from the Node Operational Credentials cluster and saves as th2_idx 
> - TH_CR1 sends the RemoveFabric command to the DUT with the FabricIndex set to th2_idx

j-ororke committed:
Creating test module TC_CADMIN_1_11:
- Test module created following matter-test-scripts task (https://github.com/project-chip/matter-test-scripts/issues/286):
-- Open commissioning window on DUT twice using ECM then BCM [DUT - Commissionee]
-- Commission DUT to TH_CR1 (can be skipped if done in a preceding test)
-- TH_CR1 reads the BasicCommissioningInfo attribute from the General Commissioning cluster on EP0 and saves the MaxCumulativeFailsafeSeconds as timeout
-- TH_CR1 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout
-- TH_CR2 fully commissions the DUT
-- TH_CR1 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout
-- TH_CR1 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout
-- TH_CR2 sends an OpenCommissioningWindow command to the DUT with the CommissioningTimeout set to timeout
-- TH_CR1 sends an RevokeCommissioning command to the DUT
-- TH_CR1 reads the FeatureMap from the Administrator Commissioning Cluster. If the feature map includes the BC feature bit, repeat steps 5-8 using the OpenBasicCommissioningWindow command
-- TH_CR2 reads the CurrentFabricIndex attribute from the Node Operational Credentials cluster and saves as th2_idx
-- TH_CR1 sends the RemoveFabric command to the DUT with the FabricIndex set to th2_idx

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Update tests.yaml

j-ororke committed:
Merge branch 'master' into test_CADMIN_1_11

j-ororke committed:
Update TC_CADMIN_1_11.py

Updated OpenbasicCommissioningWindow to update to using correct variables.

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,197 @@
+        self.step(9)
+        AC_cluster = Clusters.AdministratorCommissioning
+        features = await self.read_single_attribute(dev_ctrl=self.default_controller, node_id=self.dut_node_id,
+                                                    endpoint=0, attribute=AC_cluster.Attributes.FeatureMap)
```
Can you omit the endpoint here please? It will default to matter_test_config.endpoint, with a fallback to 0. This isn't wrong right now (administrator commissioning is always on 0), but one of the new features is going to have this cluster land on other endpoints, so we need to be able to run this on non-zero endpoints starting at SVE.

j-ororke replied:
So I did this and the code complained that the endpoint was a necessary var needed for the read_single_attribute function, so I changed to using the read_single_attribute_check_success function, however since there is only one feature for the AdministratorCommissioning cluster, if its not supported then it returns a 0 or None value, which the read_single_attribute_check_success function does not appear to like and it complains the following:
`mobly.signals.TestFailure: Details=Error reading <class 'chip.clusters.Objects.AdministratorCommissioning'>:<class 'chip.clusters.Objects.AdministratorCommissioning.Attributes.FeatureMap'> = ValueDecodeFailure(TLVValue=None, Reason=InteractionModelError(<Status.UnsupportedCluster: 195>)), Extras=None`
So due to the above I then changed to using an if else statement so if there were features on the device then it would do the read_single_attribute_check_success function, else we set the features var to the direct output from AC_cluster.Attributes.FeatureMap.value.

cecille replied:
The error there is saying that the cluster is unsupported, which means the endpoint is not 0. Can you post the TH logs there so we can check what's going on? FeatureMap should always be present and not error, so if it's missing, that's probably a sign there's something either wrong on the test or on the app.

j-ororke replied:
This appears to have been resolved with an update to both the TH and the virtual environment established in WSL on my local workstation. My apologies for the above issue mentioned.

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,197 @@
+        self.timeout = duration.maxCumulativeFailsafeSeconds
+
+        self.step(3)
+        self.count = 0
```
what's this used for?

j-ororke replied:
Sorry, that should have been removed.
I have removed it here now from the code.

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,197 @@
+
+    def steps_TC_CADMIN_1_11(self) -> list[TestStep]:
+        return [
+            TestStep(1, "Commissioning, already done", is_commissioning=True),
```
could you please add the expectations in here as well?

j-ororke replied:
Hi Cecille,
I have added the expectations to the Test steps, hopefully the ones added are acceptable.

j-ororke committed:
Merge branch 'master' into test_CADMIN_1_11

j-ororke committed:
Updated TC_CADMIN_1_11 test module:
- Removed unneccessary count variable due to no longer being needed
- Added expectations to TestSteps
- Replaced elif with else statements in OpenCommissioningWindow and OpenBasicCommissioningWindow functions
- Set busy_enum var using Clusters.AdministratorCommissioning.Enums.StatusCode.kBusy enum value
- Changed method to check features for administrator commissioning cluster to check if BC bit value is contained in features

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated TC_CADMIN_1_11:
- Removed unneeded prints from the code output

j-ororke commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,197 @@
+                nodeid=self.dut_node_id, timeout=self.timeout, iteration=10000, discriminator=self.discriminator, option=1)
+            return params
+
+        elif expectedErrCode == 0x02:
```
Hi Cecille,
I have removed the elif statement and replace it with an else statement, that way all possible error codes can be handled.

j-ororke commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,197 @@
+        await self.OpenCommissioningWindow(self.th1, 0x00)
+
+        self.step(6)
+        await self.OpenCommissioningWindow(self.th1, 0x02)
```
Hi Cecille,
Thank you for this, I have set a new var "busy_enum" which takes the value from the Clusters.AdministratorCommissioning.Enums.StatusCode.kBusy which you provided above, then uses that var in the OpenCommissioningWindow and OpenBasicCommissioningWindow functions.

restyled-commits committed:
Restyled by autopep8

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,210 @@
+
+        self.supports_bc = bool(features & AC_cluster.Bitmaps.Feature.kBasic) != 0
+
+        if self.supports_bc:
```
Do you know if the current example app supports BC? I want to ensure the test steps have all been checked by the CI / you testing.

j-ororke replied:
Hi Cecille,
After your provided guidance and support on how to use the zap tool, we were able to enable BC feature on the all-clusters app then validated that the updated method provided in the latest changes did work to open the BC commissioning window.

cecille replied:
kk, sounds like this was tested, so we're good here.

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,210 @@
+            asserts.assert_true(errcode.sdk_code == expectedErrCode,
+                                'Unexpected error code returned from CommissioningComplete')
+
+    async def OpenBasicCommissioningWindow(self, th, expectedErrCode) -> CommissioningParameters:
```
```suggestion
    async def OpenBasicCommissioningWindow(self, th: ChipDeviceCtrl, expectedErrCode: Optional[Clusters.AdministratorCommissioning.Enums.StatusCode] = None) -> CommissioningParameters:
```

cecille replied:
ditto above

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,210 @@
+                                'Unexpected error code returned from CommissioningComplete')
+
+    async def OpenBasicCommissioningWindow(self, th, expectedErrCode) -> CommissioningParameters:
+        if expectedErrCode == 0x00:
```
```suggestion
        if not expectedErrCode:
```

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,210 @@
+        if self.supports_bc:
+            self.count = 0
+            self.step("9a")
+            await self.OpenBasicCommissioningWindow(self.th1, 0x00)
```
```suggestion
            await self.OpenBasicCommissioningWindow(self.th1)
```

j-ororke committed:
Apply suggestions from code review

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Update TC_CADMIN_1_11.py

-Added Optional function from typing library to resolve issue with latest code change

j-ororke committed:
Merge branch 'master' into test_CADMIN_1_11

j-ororke committed:
Update TC_CADMIN_1_11.py

- Due to recent changes, had to update to using PyChipError for ctx.exception.err in order to handle error raised during commissioning properly

j-ororke committed:
Updating TC_CADMIN_1_11:
- Updated method for OpenBasicCommissiongWindow with BC feature enabled

restyled-commits committed:
Restyled by autopep8

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,218 @@
+                                'Unexpected error code returned from CommissioningComplete')
+
+    async def read_currentfabricindex(self, th: ChipDeviceCtrl) -> int:
+        cluster = Clusters.Objects.OperationalCredentials
```
```suggestion
        cluster = Clusters.OperationalCredentials
```

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,218 @@
+                     "TH_CR1 removes TH_CR2 fabric using th2_idx"),
+        ]
+
+    def generate_unique_random_value(self, value):
```
I don't think this function is used anywhere

j-ororke replied:
Thank you, i have removed this function from this test script as its not needed here.

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,218 @@
+        self.discriminator = random.randint(0, 4095)
+        th2_certificate_authority = self.certificate_authority_manager.NewCertificateAuthority()
+        th2_fabric_admin = th2_certificate_authority.NewFabricAdmin(vendorId=0xFFF1, fabricId=self.th1.fabricId + 1)
+        self.th2 = th2_fabric_admin.NewController(nodeId=2, useTestCommissioner=True)
```
you don't need the test commissioner. That's an override for testing commissioning failures.

j-ororke replied:
Thank you, I have removed the useTestCommissioner arg when establishing the self.th2 controller.

cecille commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,218 @@
+
+class TC_CADMIN_1_11(MatterBaseTest):
+    async def OpenCommissioningWindow(self, th, expectedErrCode) -> CommissioningParameters:
+        if expectedErrCode == 0x00:
```
Can you use none instead of 0 here?

j-ororke replied:
No problem, I have updated this to be None instead of 0 in the script now

cecille reviewed (approved):
Couple of minor things, but this looks good once those are sorted.

j-ororke committed:
Update src/python_testing/TC_CADMIN_1_11.py

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Merge branch 'master' into test_CADMIN_1_11

j-ororke committed:
Updated TC_CADMIN_1_11 test module:
- Removed useTestCommissioner when creating self.th2
- Changed to using None instead of 0 as variable value for OpenCommissioningWindow()
- Removed generate_unique_randow_value() as not needed in this test module

j-ororke committed:
Updating TC_CADMIN_1_11 test module:
- Resolved linting error

j-ororke committed:
Merge branch 'master' into test_CADMIN_1_11

j-ororke committed:
Updated TC_CADMIN_1_11:
- Updated due to latest changes to matter_testing support module

restyled-commits committed:
Restyled by autopep8

raul-marquez-csa commented:
The YAML script for the test `src/app/tests/suites/certification/Test_TC_CADMIN_1_11.yaml` is still present, it should be removed as well as references to it, correct?

j-ororke committed:
Removed TC_CADMIN_1_11.yaml and updated TC_CADMIN_1_11:
- Removed Test_TC_CADMIN_1_11.yaml
- Updated TC_CADMIN_1_11 python test module to enable verbose output in CI/CD pipeline.

raul-marquez-csa commented on src/python_testing/TC_CADMIN_1_11.py:
```diff
@@ -0,0 +1,213 @@
+# test-runner-run/run1/quiet: True
+# test-runner-run/run1/app-args: --discriminator 1234 --KVS kvs1 --trace-to json:${TRACE_APP}.json
+# test-runner-run/run1/script-args:  --storage-path admin_storage.json --commissioning-method on-network --discriminator 1234 --passcode 20202021 --PICS src/app/tests/suites/certification/ci-pics-values --trace-to json:${TRACE_TEST_JSON}.json --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
+# === END CI TEST ARGUMENTS ===
```
The nomenclature for the CI test arguments has been updated, they look something like this
```
# === BEGIN CI TEST ARGUMENTS ===
# test-runner-runs:
#   run1:
#     app: ${ALL_CLUSTERS_APP}
#     app-args: --discriminator 1234 --KVS kvs1 --trace-to json:${TRACE_APP}.json
#     script-args: >
#       --storage-path admin_storage.json
#       --commissioning-method on-network
#       --discriminator 1234
#       --passcode 20202021
#       --trace-to json:${TRACE_TEST_JSON}.json
#       --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
#     factory-reset: true
#     quiet: true
# === END CI TEST ARGUMENTS ===
```
You can check out most python scripts to match it to your needs.

j-ororke replied:
Thank you, Raul, I had forgotten to update that to match the latest version for CI/CD arguments.

j-ororke committed:
Updated TC_CADMIN_1_11 test module:
- Updated CI/CD nominclature for this test to match latest version

raul-marquez-csa commented:
This script is referenced in `src/app/tests/suites/manualTests.json` 

https://github.com/j-ororke/connectedhomeip/blob/1a2c7a9cce0b65ae27877303ca9ec2fe6b3b4a22/src/app/tests/suites/manualTests.json#L169-L174

I believe it no longer belongs there.

j-ororke committed:
Updated src/app/tests/suites/manualTests.json:
- Removed Test_TC_CADMIN_1_11 from MultipleFabrics section
